espnet2/gan_codec/speechtokenizer/speechtokenizer.py

```diff
@@ -0,0 +1,801 @@
+        bw = self.target_bandwidths[random.randint(0, max_idx)]
+        
+        # Forward quantizer
+        quantized, _, _, commit_loss, quantized_list = self.quantizer(encoder_out, self.frame_rate, bw, layers=layers)
+        
+        feature = rearrange(quantized_list[0], 'b d t -> b t d')
+        feature = self.transform(feature)    
+        
+        quantization_loss = self.l1_quantization_loss(
+            encoder_out, quantized.detach()
+        ) + self.l2_quantization_loss(encoder_out, quantized.detach())
+
+        resyn_audio = self.decoder(quantized)
+
+        if use_dual_decoder:
+            resyn_audio_real = self.decoder(encoder_out)
+        else:
+            resyn_audio_real = None
+        return resyn_audio, commit_loss, quantization_loss, resyn_audio_real, feature # return the feature
+
+    def forward_feature(self, 
+                            x: torch.Tensor, 
+                            layers: list=None):
+            '''
+
+            Parameters
+            ----------
+            x : torch.tensor
+                Input wavs. Shape should be (batch, channels, timesteps).
+            layers : list[int], optional
+                Layers of RVQ should return quantized result. The default is all layers.
+
+            Returns
+            -------
+            quantized_list : list[torch.tensor]
+                Quantized of required layers.
+
+            '''
+            e = self.encoder(x)
+            layers = layers if layers else list(range(self.n_q))
+            max_idx = len(self.target_bandwidths) - 1
+            # randomly pick up one bandwidth
+            bw = self.target_bandwidths[random.randint(0, max_idx)]
+            # Forward quantizer
+            quantized, _, _, commit_loss, quantized_list = self.quantizer(e, self.frame_rate, bw, layers=layers)
+            return quantized_list
+    
+    def encode(
+        self,
+        x: torch.Tensor,
+        target_bw: Optional[float] = None,
+    ):
+        """SpeechTokenizer encoding.
+
+        Args:
+            x (torch.Tensor): Input tensor of shape (B, 1, T).
+        Returns:
+            torch.Tensor: neural codecs in shape (). #######
+        """
+        if x.dim() == 1:
+            x = x.view(1, 1, -1)
+        elif x.dim() == 2:
+            x = x.unsqueeze(1)
+
+        encoder_out = self.encoder(x)
+        if target_bw is None:
+            bw = self.target_bandwidths[-1]
+        else:
+            bw = target_bw
+        codes = self.quantizer.encode(encoder_out, self.frame_rate, bw)
+        return codes
+
+    def decode(self, codes: torch.Tensor):
+        """SpeechTokenizer decoding.
+
+        Args:
+            codecs (torch.Tensor): neural codecs in shape ().
+        Returns:
+            torch.Tensor: resynthesized audio.
+        """
+        quantized = self.quantizer.decode(codes)
+        resyn_audio = self.decoder(quantized)
+        return resyn_audio
+
+
+class SpeechTokenizerDiscriminator(nn.Module):
```

Comment:
You seem to resue the soundstream discriminator but it does not match the speech tokenizer paper. Could you please double check?